nativelib/src/main/resources/scala-native/zone/MemoryPool.c

```diff
@@ -0,0 +1,128 @@
+#else // Unix
+    if (chunk->start == MAP_FAILED) {
+        fprintf(stderr, "Failed to allocate memory chunk of size %zu\n",
+                chunk->size);
+    }
+#endif
+    chunk->next = pool->chunk;
+    pool->chunk = chunk;
+}
+
+void memorypool_alloc_page(MemoryPool *pool, size_t size) {
+    if (pool->chunk->offset >= pool->chunk->size ||
+        pool->chunk->offset + size > pool->chunk->size) {
+        memorypool_alloc_chunk(pool, max(pool->chunkSize, size));
+    }
+    MemoryPage *page = malloc(sizeof(MemoryPage));
+    page->start = pool->chunk->start + pool->chunk->offset;
+    page->offset = 0;
+    page->size = size;
+    page->next = pool->page;
+    pool->chunk->offset += page->size;
+    pool->page = page;
+}
+
+MemoryPage *memorypool_claim(void *pool) {
+    return memorypool_claim_with_min_size(pool, 0);
+}
+
+/** Borrow a single unused page, to be reclaimed later. */
+MemoryPage *memorypool_claim_with_min_size(void *_pool, size_t min_size) {
+    MemoryPool *pool = (MemoryPool *)_pool;
+    size_t page_size = MEMORYPOOL_PAGE_SIZE;
+    if (min_size > MEMORYPOOL_MAX_CHUNK_SIZE) {
+        fprintf(stderr, "Requested size is too large: %zu\n", min_size);
+        return NULL;
+    }
+    while (min_size > page_size) {
+        page_size *= 2;
+        if (page_size > MEMORYPOOL_MAX_CHUNK_SIZE) {
+            page_size = MEMORYPOOL_MAX_CHUNK_SIZE;
+        }
+    }
+    if (pool->page == NULL) {
+        memorypool_alloc_page(pool, page_size);
+    }
+    MemoryPage *result = pool->page;
+    pool->page = result->next;
+    result->next = NULL;
+    result->offset = 0;
+    // Notify the GC that the page is in use.
+    scalanative_add_roots(result->start, result->start + result->size);
+    return result;
+}
+
+/** Reclaimed a list of previously borrowed pages. */
+void memorypool_reclaim(void *_pool, void *_head_page, void *_tail_page) {
+    MemoryPool *pool = (MemoryPool *)_pool;
+    MemoryPage *head = (MemoryPage *)_head_page;
+    MemoryPage *tail = (MemoryPage *)_tail_page;
+    // Notify the GC that the pages are no longer in use.
+    MemoryPage *page = head;
+    while (page != NULL) {
+        scalanative_remove_roots(page->start, page->start + page->size);
+        if (page == tail)
+            break;
+        page = page->next;
+    }
+    // Append the reclaimed pages to the pool.
+    tail->next = pool->page;
+    pool->page = head;
+}
+
+void memorypool_free(void *_pool) {
+    MemoryPool *pool = (MemoryPool *)_pool;
+    // Free chunks.
+    MemoryChunk *chunk = pool->chunk, *pre_chunk = NULL;
+    while (chunk != NULL) {
+        pre_chunk = chunk;
+        chunk = chunk->next;
+        memoryUnmap(pre_chunk->start, pre_chunk->size);
+        free(pre_chunk);
+    }
+    // Free pages.
+    MemoryPage *page = pool->page, *pre_page = NULL;
+    while (page != NULL) {
+        pre_page = page;
+        page = page->next;
+        free(pre_page);
+    }
+    // Free the pool.
+    free(pool);
+}
```

Comment:
There is plenty of files without a blank line at the EOF (warned by GitHub with red sign that can be seen here). In most cases its not a problem, but it's a good practice to add this spare blank line, otherwise some legacy programs might have some issues, e.g. git pathes AFAIK